packages/contracts-bedrock/src/libraries/Hashing.sol

```diff
@@ -121,4 +121,28 @@ library Hashing {
             )
         );
     }
+
+    /// @notice Generates a unique hash for a message to be relayed across chains. This hash is
+    ///         used to identify the message and ensure it is not relayed more than once.
+    /// @param _destination Chain ID of the destination chain.
+    /// @param _source Chain ID of the source chain.
+    /// @param _nonce Unique nonce associated with the message to prevent replay attacks.
+    /// @param _sender Address of the user who originally sent the message.
+    /// @param _target Address of the contract or wallet that the message is targeting on the destination chain.
+    /// @param _message The message payload to be relayed to the target on the destination chain.
+    /// @return Hash of the encoded message parameters, used to uniquely identify the message.
+    function hashL2toL2CrossDomainMessengerRelayMessage(
+        uint256 _destination,
+        uint256 _source,
+        uint256 _nonce,
+        address _sender,
+        address _target,
+        bytes memory _message
+    )
+        internal
+        pure
+        returns (bytes32)
+    {
+        return keccak256(abi.encode(_destination, _source, _nonce, _sender, _target, _message));
+    }
```

Comment:
how is this library abstraction necessary?
would it be ok to remove it? this library could get a bit crazy LOC-wise

Reply:
this could be useful if an app ever needs to generate the message hash outside of L2CDM. For example, a contract needs to check a message status for a contract they do not own that calls into `L2CDM.sendMessage`, but that contract does not pass up the message hash.

Reply:
I see, my fear is that either-way the developers would need to go and grab that specific piece of code, and having a library that would hold them all can get out of hand quite quickly.
do you think this could be an internal function? just to have the pretty natspec and allow devs to easily copy it over

Reply:
initially I did have it as an internal function but per @tynes comment: https://github.com/ethereum-optimism/optimism/pull/12089#discussion_r1774021322 I moved it into Hashing.sol. I would prefer either Hashing or internal over doing this inline because it is important that this does not diverge from the hash generated in relayMessage

Reply:
ah rip, mark front-ran me to both changes.
@tynes what do you think of my counter-argument above?
I rather not have a huge library with random hashing functions.
we could ask devs what they think, personally I rather not have to dig into a library, but if I had to, i'd probably just copy the function over.